### PR TITLE
KAFKA-17545: Removing process fetch queue

### PR DESCRIPTION
**Pull request by apoorvmittal10**
The PR removed the process fetch queue as we have moved to share fetch purgatory. As the fetches are passed to purgatory hence close of SharePartitionManager can't validate if all fetches are completed on shutdown hence removed the test case. We shall add a test case in ReplicaManager which should guarantee that all share fetches are completed on shutdown.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by adixitconfluent — APPROVED** [review]
Thanks for the PR. LGTM.

**Comment by apoorvmittal10** [comment]
Future tasks associated with refactor:

[KAFKA-17829](https://issues.apache.org/jira/browse/KAFKA-17829) - We need to add an integration test which verifies that on shutdown of the delayed share fetch purgatory, the share fetch requests which are present inside the purgatory return with an erroneous future.

File: core/src/main/java/kafka/server/share/SharePartitionManager.java [hunk 1]
```diff
@@ -184,30 +172,28 @@ private SharePartitionManager(
...
-        this.defaultRecordLockDurationMs = defaultRecordLockDurationMs;
-        this.timer = new SystemTimerReaper("share-group-lock-timeout-reaper",
-            new SystemTimer("share-group-lock-timeout"));
-        this.maxDeliveryCount = maxDeliveryCount;
-        this.maxInFlightMessages = maxInFlightMessages;
-        this.persister = persister;
-        this.groupConfigManager = groupConfigManager;
-        this.shareGroupMetrics = new ShareGroupMetrics(Objects.requireNonNull(metrics), time);
+        this(
+            replicaManager,
```
Comment (junrao) [inline core/src/main/java/kafka/server/share/SharePartitionManager.java]: merge with previous line?

**Review by junrao — COMMENTED** [review]
@apoorvmittal10 : Thanks for the PR. Just a minor comment.

File: core/src/main/java/kafka/server/share/SharePartitionManager.java [hunk 2]
```diff
@@ -547,31 +522,11 @@ private static String partitionsToLogString(Collection<TopicIdPartition> partiti
...
-
-        ShareFetchData shareFetchData = fetchQueue.poll();
-        if (shareFetchData == null) {
-            // No more requests to process, so release the lock. Though we should not reach here as the lock
-            // is acquired only when there are requests in the queue. But still, it's safe to release the lock.
-            releaseProcessFetchQueueLock();
-            return;
-        }
-
+    void processShareFetch(ShareFetchData shareFetchData) {
```
Comment (junrao) [inline core/src/main/java/kafka/server/share/SharePartitionManager.java:525]: While you are at this class, in the following code, `return` seems unnecessary. 

```
                sharePartition.maybeInitialize().whenComplete((result, throwable) -> {
                    if (throwable != null) {
                        maybeCompleteInitializationWithException(sharePartitionKey, shareFetchData.future(), throwable);
                        return;
                    }
```

File: core/src/main/java/kafka/server/share/SharePartitionManager.java (hunk 1 above)
Comment (apoorvmittal10) [reply core/src/main/java/kafka/server/share/SharePartitionManager.java]: Done.

File: core/src/main/java/kafka/server/share/SharePartitionManager.java (hunk 2 above)
Comment (apoorvmittal10) [reply core/src/main/java/kafka/server/share/SharePartitionManager.java:525]: Done. Though I am reafactoring the initilization handling in Share Partition Manager as we need to address the concern when Share Partition is loaded late and request is added to purgatory, etc. The PR will be out soon.

**Review by junrao — APPROVED** [review]
@apoorvmittal10 : Thanks for the updated PR. LGTM

File: core/src/main/java/kafka/server/share/SharePartitionManager.java [hunk 3]
```diff
@@ -611,20 +564,9 @@ void maybeProcessFetchQueue() {
...
-            // Release the lock so that other threads can process the queue.
-            releaseProcessFetchQueueLock();
-            // If there are more requests in the queue, then process them.
-            if (!fetchQueue.isEmpty())
-                maybeProcessFetchQueue();
-
         } catch (Exception e) {
             // In case exception occurs then release the locks so queue can be further processed.
             log.error("Error processing fetch queue for share partitions", e);
-            releaseProcessFetchQueueLock();
```
Comment (junrao) [inline core/src/main/java/kafka/server/share/SharePartitionManager.java:624]: Could we remove the comment in the catch clause? Also, is the code really throwing any exception? If not, we can just remove the catch clause.

**Review by junrao — COMMENTED** [review]
@apoorvmittal10 : Left another comment.

File: core/src/main/java/kafka/server/share/SharePartitionManager.java (hunk 3 above)
Comment (apoorvmittal10) [reply core/src/main/java/kafka/server/share/SharePartitionManager.java:624]: Sure @junrao, I am adding an additional handling where the catch block is required. I ll send the changes in open partition leader PR: https://github.com/apache/kafka/pull/16842

Comment (junrao) [reply core/src/main/java/kafka/server/share/SharePartitionManager.java:624]: Thanks. Also, `SharePartition.maybeAcquireFetchLock` still has comments on fetch queue.
```
    /**
     * Prior to fetching records from the leader, the fetch lock is acquired to ensure that the same
     * share partition does not enter a fetch queue while another one is being fetched within the queue.
     * The fetch lock is released once the records are fetched from the leader.
     *

```

**Pull request merged**